[addons] Thick cleaning up of the C++ API

AlwinEsch opened the pull request:
## Description

Since the following requests make it necessary to re-create all add-ons (not absolutely necessary with regard to C++17), I have made a few changes here to clean up, partly to get them in the sandbox (not directly in contact with it) and to have it prepared for a feature on PVR addons to have multiple instances.

- https://github.com/xbmc/xbmc/pull/20705
- https://github.com/xbmc/xbmc/pull/20699
- https://github.com/AlwinEsch/kodi/pull/46

For a review, it is better to use the individual commits, as they are only indirectly related and could also come independently (only all breaks the API :smirk:).

I still have to change and test all addons, so I can't say 100% whether it's running correctly, especially with game addons because there it doesn't have instance support.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons] Remove fallbacks to old

In the following changes, a min version increase is necessary, so it makes sense to remove this now because all add-ons need an update anyway.

AlwinEsch committed:
[addons] remove not used function values on "ADDON_Create"

In ADDON_Create there were previously two values which are not used, they are removed because a MIN version increase is required.

Part of the previous cleanup of the API.

AlwinEsch committed:
[addons] remove "ADDON_STATUS (*get_status)();"

This has not been used for some time and since the following changes result in a MIN version increase, it can be removed.

AlwinEsch committed:
[addons] cleanup ADDONCREATOR macro

With the addition of "using namespace" in function, the code becomes cleaner because the line breaks are no longer necessary.

AlwinEsch committed:
[addons] Pass handling pointers in function values

The important reason for this is to have "C" ABI more correctly, where its handling pointer is passed in every function call.

Previously this was stored in a structure on the add-on side, which is not really clean.
It is thought to have cleaner add-ons that are purely based on "C".

There will also be some commits to make it even more correct.

AlwinEsch committed:
[addons] use typedef's about function pointers

With this, the addon function pointers are set in the structure using typedefs, this makes it cleaner and easier to use in other places in the planned sandbox.

These typedef's are only used for calls from Kodi to addon, in the other way addon to Kodi they stay in the old way.

AlwinEsch committed:
[addons] Use the handover setting for each type of different function

Previously, it was transferred to all types in one function using a `void*`, but since such use can be very dangerous and it does not work in the sandbox, it is divided into separate functions here.
On the C++ side it remains as before, since it is easier to handle in a class.

AlwinEsch committed:
[addons] remove some structure definitions

In the correct "C" ABI these are not necessary, since a `struct` must already be at begin at the positions used. Hence, it can be removed and make the code cleaner.

AlwinEsch committed:
[addons] remove unused string "kodi_base_api_version" in API

The string value kodi_base_api_version stored in the "C" structure is never used and is hereby removed as an adjustment.

AlwinEsch committed:
[addons] Add setting functions in instance class

With this, own functions are inserted independently of "kodi::addon" in order to call up settings in Kodi using the IAddonInstanceHandler class.
At the moment they still use the same functions as the functions in "kodi::addon::" using the "settings.xml" of the addon.
In the near future these new functions will be revised (only in Kodi's side) to allow addon instances to have their own settings (e.g. in PVR to use several backends in one addon).

The background is when this new support is introduced into Kodi, to avoid that all addons need an update again.

AlwinEsch committed:
[addons] Allow instance class to retrieve a user path

At the moment this call still returns the conventional add-on user class, in future changes the instance will be assigned its own user path, e.g. in the case of planned PVR add-ons which can connect different backends using different settings.

As a note on this and previous commits, the add-on documentation will be introduced and revised in a separate request. To do it in these commits would take a long time to get correct.

AlwinEsch committed:
[addons] increase ADDON_GLOBAL_VERSION_MAIN[_MIN] to 2.0.0

Related to changes before.

WARNING: All addons need a update about!

ksooo reviewed (commented):
Awesome. Just a question, see my review comment

AlwinEsch commented on xbmc/addons/kodi-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -337,6 +337,7 @@ class ATTR_DLL_LOCAL CAddonBase
   ///
   CAddonBase()
   {
+    CPrivateBase::m_interface->toAddon->create = nullptr;
```
Is complete unused on C++, only added to have in "C" ABI based addon case available, if becomes needed.
Was come to it as a `destroy` in structure, but no `create` and if a "C" based class handling comes somewhere, it looks more clean :sweat_smile:.

ksooo replied:
Okay then, I was just wondering whether this is intended.

AlwinEsch committed:
[addons][pvr] change and move ADDON_HANDLE to PVR_HANDLE

The ADDON_HANDLE is only used in PVR, since it should not be used any longer, it has now been moved to "kodi/c-api/addon-instance/pvr/pvr_defines.h" and renamed to PVR_HANDLE.

AlwinEsch committed:
[addons] Use KODI_ADDON_BACKEND_HDL in AddonBase.h

Previously the values were set as `void*`, since this is a bit ugly, it is replaced here with `const KODI_ADDON_BACKEND_HDL`.
In principle, this could be done at any point, but it would extremely blown the request. Since this can also be done separately (without addon updates), it is only related to AddonBase.h here.

AlwinEsch committed:
[addons] revised C++ kodi::[addon] API function places

This revises the places where functions are located, so some of the things are moved from General.h to AddonBase.h and these then use the namespace "kodi :: addon", in the background it is more suitable to the Python addon system, where the here in the C++ moved, already in Python at self.Addon.

In addition, the new structure AddonToKodiFuncTable_kodi_addon has been added to make changes easier and not have to tackle the MIN version in the future.

Unfortunately, the parts in headers had to be moved a bit, which also made it a bit bigger in commit.

AlwinEsch commented:
About audiodecoder & audioencoder addons are requests there and was the functional check on my Linux OK. 
Tested by C++17, the VFS change and my changes-

I'm tommorow a bit away about stupid aircraft works, hope to continue the other addons a bit later on evening.

phunkyfish reviewed (commented):
Nice, it looks good to me. Will there be other dependent PRs or should this one fully prepare things for the sandbox?

AlwinEsch commented:
> Nice, it looks good to me. Will there be other dependent PRs or should this one fully prepare things for the sandbox?

No, about sandbox has this only indirect todo, are few parts there where I used also on sandbox, but they not direct relates to it.
For here mostly to @ksooo wanted support about multiple instances by PVR.

AlwinEsch commented on xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon_base.h:
```diff
@@ -249,24 +249,33 @@ extern "C"
+                                                                  KODI_HANDLE instance);
+  typedef ADDON_STATUS(ATTR_APIENTRYP PFN_KODI_ADDON_SET_SETTING_V1)(const KODI_ADDON_HDL hdl,
+                                                                     const char* settingName,
+                                                                     const void* settingValue);
```
The add about the typedef's is the most needed part within my sandbox, as this becomes several times used on the autogeneration to create the library for it.

The reason why these are only for calls from Kodi to add-ons has the background that the add-on has to set its pointers there.

In the case of callbacks from add-ons to Kodi, it will be stored in the sandbox as correct functions so that it can be used very easily and cleanly in "C" or as a basis for other languages, see https://github.com/AlwinEsch/kodi/blob/new-interface-2/xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon_base.h#L183-L249.

garbear commented:
I was confused what "ifc" stood for, but I think it's "instance function callback". Could the abbreviation be lengthened, or the full term documented in multiple places?

AlwinEsch commented:
IFC was the short about "interface".

Is about parts where works between Kodi and addon.

AlwinEsch committed:
[addons] rename enum AddonLog to ADDON_LOG

The background is to be more in the "C" format style.

AlwinEsch committed:
[addons] Revise addon "C" structures

This makes a big change, with which the instance function structures are stored in Kodi's `IAddonInstanceHandler` class and a structure with union's instead of "void*" is transferred in" C ". The background is easier to expand and can also be used on the add-on side when creating an instance. The added class `IInstanceInfo` can then be used to query the values.
This makes it easier for extensions because the values in the CreateInstance function do not always have to be changed and updates with changes are required for all add-ons.

AlwinEsch commented:
The orgy is through so far, apart from image decoder addons are all now with requests.

I update to imagedecoder addons after this one and the other one (https://github.com/xbmc/xbmc/pull/20508) in Kodi. Otherwise it would be double the work.

The tests so far were OK on my Linux.
Can you then take a look and give your OK or comments?

garbear reviewed (approved):
Game and Peripheral changes look good. I don't find `m_ifc` to be a very clear name, so maybe `m_iface` or `m_interface`?

AlwinEsch commented:
Also done few tests within Windows and was working too.

About `m_ifc`, I leave as is, is better for me, as the name only secondary and should by short as possible.
This is e.g. it will come on sandbox project:
```cpp
int CAudioEncoder::Encode(int nNumBytesRead, uint8_t* pbtStream)
{
  return m_ifc->kodi_addoninstance_audioencoder_h->kodi_addon_audioencoder_encode_v1(m_addonInstance, pbtStream, nNumBytesRead);
}
```
The names are autogenerated and as `kodi_addoninstance_audioencoder_h` to part in reference to related header and `kodi_addon_audioencoder_encode_v1` the name of related API function.

This both are they primary parts to know and the `m_ifc` only to have as something available.

AlwinEsch commented:
Can someone else look over the request and approve, want to make sure that no bad part is overseen.
